middleware/oauth2/oauth2_middleware.go

```diff
@@ -0,0 +1,101 @@
+// ------------------------------------------------------------
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+// ------------------------------------------------------------
+
+package oauth2
+
+import (
+	"encoding/json"
+	"strings"
+
+	"github.com/dapr/components-contrib/middleware"
+	"github.com/google/uuid"
+	"github.com/valyala/fasthttp"
+	"golang.org/x/oauth2"
+	"context"	
+)
+
+// Metadata is the oAuth middleware config
+type oAuth2MiddlewareMetadata struct {
+	ClientID       string `json:"clientID"`
+	ClientSecret   string `json:"clientSecret"`
+	Scopes         string `json:"scopes"`
+	AuthURL        string `json:"authURL"`
+	TokenURL       string `json:"tokenURL"`
+	AuthHeaderName string `json:"authHeaderName"`
+	RedirectURL    string `json:"redirectURL"`
+}
+
+// NewOAuth2Middleware returns a new oAuth2 middleware
+func NewOAuth2Middleware() *Middleware {
+	return &Middleware{}
+}
+
+// Middleware is an oAuth2 authentication middleware
+type Middleware struct {
+}
+
+// GetHandler retruns the HTTP handler provided by the middleware
+func (m *Middleware) GetHandler(metadata middleware.Metadata) (func(h fasthttp.RequestHandler) fasthttp.RequestHandler, error) {
+	meta, err := m.getNativeMetadata(metadata)
+	if err != nil {
+		return nil, err
+	}
+
+	return func(h fasthttp.RequestHandler) fasthttp.RequestHandler {
+		return func(ctx *fasthttp.RequestCtx) {
+			conf := &oauth2.Config{
+				ClientID:     meta.ClientID,
+				ClientSecret: meta.ClientSecret,
+				Scopes:       strings.Split(meta.Scopes, ","),
+				RedirectURL:  meta.RedirectURL,
+				Endpoint: oauth2.Endpoint{
+					AuthURL:  meta.AuthURL,
+					TokenURL: meta.TokenURL,
+				},
+			}
+			state := string(ctx.FormValue("state"))
+			if state == "" {
+				id, _ := uuid.NewUUID()
+				cookie := fasthttp.Cookie{}
+				cookie.SetKey("state")

```

Comment:
Can you make a constant for "state"? also for "code"

Reply:
included in new code